src/AdvancedCircuitBreaker.php

```diff
@@ -0,0 +1,173 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\CircuitBreaker;
+
+use PrestaShop\CircuitBreaker\Contract\ClientInterface;
+use PrestaShop\CircuitBreaker\Contract\StorageInterface;
+use PrestaShop\CircuitBreaker\Contract\SystemInterface;
+use PrestaShop\CircuitBreaker\Contract\TransitionDispatcherInterface;
+use PrestaShop\CircuitBreaker\Exception\UnavailableServiceException;
+
+/**
+ * This implementation of the CircuitBreaker is a bit more advanced than the SimpleCircuitBreaker,
+ * it allows you to setup your client, system, storage and dispatcher.
+ */
+class AdvancedCircuitBreaker extends PartialCircuitBreaker
+{
+    /** @var TransitionDispatcherInterface */
+    protected $dispatcher;
+
+    /** @var callable|null */
+    protected $defaultFallback;
+
+    /**
+     * @param SystemInterface $system
+     * @param ClientInterface $client
+     * @param StorageInterface $storage
+     * @param TransitionDispatcherInterface $dispatcher
+     */
+    public function __construct(SystemInterface $system, ClientInterface $client, StorageInterface $storage, TransitionDispatcherInterface $dispatcher)
+    {
+        parent::__construct($system, $client, $storage);
+        $this->dispatcher = $dispatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function call(
+        $service,
+        array $serviceParameters = [],
+        callable $fallback = null
+    ) {
+        $transaction = $this->initTransaction($service);
+
+        try {
+            if ($this->isOpened()) {
+                if (!$this->canAccessService($transaction)) {
+                    return $this->callFallback($fallback);
+                }
+
+                $this->moveStateTo(State::HALF_OPEN_STATE, $service);
+                $this->dispatchTransition(
+                    Transition::CHECKING_AVAILABILITY_TRANSITION,
+                    $service,
+                    $serviceParameters
+                );
+            }
+
+            $response = $this->request($service, $serviceParameters);
+            $this->moveStateTo(State::CLOSED_STATE, $service);
+            $this->dispatchTransition(
+                Transition::CLOSING_TRANSITION,
+                $service,
+                $serviceParameters
+            );
+
+            return $response;
+        } catch (UnavailableServiceException $exception) {
+            $transaction->incrementFailures();
+            $this->storage->saveTransaction($service, $transaction);
+            if (!$this->isAllowedToRetry($transaction)) {
+                $this->moveStateTo(State::OPEN_STATE, $service);
+                $transition = Transition::OPENING_TRANSITION;
+                if ($this->isHalfOpened()) {
+                    $transition = Transition::REOPENING_TRANSITION;
+                }
+                $this->dispatchTransition($transition, $service, $serviceParameters);
+
+                return $this->callFallback($fallback);
```

Comment:
What if you don't have callback here and no default callback?

Reply:
Then the `callFallback` from `PartialCircuitBreaker` will detect it and return `''`